Refactor targeting pack downloads, and framework version selection

dsplaisted opened the pull request:
This is currently WIP, I'm sending the PR now to get CI results.

This includes the following:

- Refactors FrameworkReference and targeting pack resolution so that all applicable targeting packs for the target framework will be downloaded.  This is required to support dotnet/cli#10666 (once NuGet/Home#7342 is implemented)
- Fixes #2792.  The following are the ways (in priority order) that the "runtime framework version" for a shared framework can be controlled:
    - `RuntimeFrameworkVersion` metadata on the `FrameworkReference` item
    - `RuntimeFrameworkVersion` MSBuild property (applies to all shared frameworks)
    - Then, use either the `LatestRuntimeFrameworkVersion` or the `DefaultRuntimeFrameworkVersion` of the `KnownFrameworkReference`, based on
        - The value (if set) of `TargetLatestRuntimePatch` metadata on the `FrameworkReference`
        - The `TargetLatestRuntimePatch` MSBuild property (which defaults to True if `SelfContained` is true, and False otherwise)

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveRuntimePackAssets.cs:
```diff
@@ -12,15 +12,25 @@ public class ResolveRuntimePackAssets : TaskBase
 
         protected override void ExecuteCore()
         {
             List<TaskItem> runtimePackAssets = new List<TaskItem>();
```
suggestion: extract some methods or even classes. This method is too long. Too many nested layer

dsplaisted committed:
Refactor targeting pack downloads, and support more options for selecting runtime version

dsplaisted committed:
Set test run title to distinguish between different runs

dsplaisted commented:
@dotnet/dotnet-cli @nguerrera @wli3 These code changes are now ready for review.  I still need to look at adding more test coverage.

nguerrera commented:
(Reading description, not code yet...)

Does this mean that there will have to be a targeting pack package produced for every servicing event, even if there is no API change?

dsplaisted commented:
The `RuntimeFrameworkVersion` (either as metadata or a property) isn't used for the targeting pack version.  You can set the `TargetingPackVersion` metadata on a `FrameworkReference` to specify the targeting pack version (with these changes).

dsplaisted committed:
Support disabling targeting pack download via EnableTargetingPackDownload property

Fixes #2875

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs:
```diff
@@ -24,51 +31,71 @@ public class ResolveTargetingPackAssets : TaskBase
             }
 
+            HashSet<string> frameworkReferenceNames = new HashSet<string>(FrameworkReferences.Select(fr => fr.ItemSpec), StringComparer.OrdinalIgnoreCase);
+            RuntimeFrameworksToRemove = RuntimeFrameworks.Where(rf => !frameworkReferenceNames.Contains(rf.GetMetadata(MetadataKeys.FrameworkName)))
```
just question, you need to remove RuntimeFrameworks?

dsplaisted replied:
ResolveFrameworkReferences now resolves all the possible shared frameworks that could be referenced.  This is because eventually we will support [transitive shared framework references](https://github.com/dotnet/cli/issues/10666).  It won't be until we run ResolvePackageAssets that we know for sure which ones will be referenced, so we need to filter the targeting packs, runtime packs, and RuntimeFrameworks after that target runs.

dsplaisted committed:
Refactor into calling task once to resolve app hosts

dsplaisted committed:
Only add KnownAppHostPack item from SDK if it wasn't added in bundled versions from core-sdk

dsplaisted committed:
Remove TODO

dsplaisted committed:
Fix test

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveAppHosts.cs:
```diff
@@ -0,0 +1,166 @@
+
+            if (knownAppHostPacksForTargetFramework.Count > 1)
+            {
+                throw new InvalidOperationException("Multiple KnownAppHostPack items applied to the specified target framework, which is not supposed to happen");
```
I'm torn on this being a non-BuildErrorException because KnownAppHostPack could be modified by anyone. Unlikely though, and I guess this is just moved from elsewhere.

dsplaisted replied:
I feel like it's OK to throw these types of errors when the invalid input is something that wouldn't be public if there was a concept of non-public items in MSBuild.

nguerrera replied:
Should we put "_" in front of it if it wouldn't be public?

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs:
```diff
@@ -55,167 +51,173 @@ public class ResolveFrameworkReferences : TaskBase
+                }
+                targetingPack.SetMetadata(MetadataKeys.PackageVersion, targetingPackVersion);
+
+                targetingPack.SetMetadata(MetadataKeys.RelativePath, "");
```
Why do we need an explicit empty metadata value? Is this so that it can become "data" folder later, and also not the same place for all targeting packs all at once?

dsplaisted replied:
I think it was for parity with the runtime and apphost packs.  However, the asset resolution has changed since then so I can remove it.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.TargetingPackResolution.targets:
```diff
@@ -38,6 +39,15 @@ Copyright (c) .NET Foundation. All rights reserved.
+                        AppHostPackNamePattern="runtime.**RID**.Microsoft.NETCore.DotNetAppHost"
+                        AppHostPackVersion="3.0.0-preview-27218-01"
+                        AppHostRuntimeIdentifiers="freebsd-x64;linux-arm;linux-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;tizen.4.0.0-armel;tizen.5.0.0-armel;win-arm;win-arm64;win-x64;win-x86"
+                        />
```
Didn't you do this already?

dsplaisted replied:
Stage 0 needed to be updated, now done.

nguerrera commented:
Do we have a read on the performance cost of these new Resolve tasks? I am holding back on commenting on all the Linq and what not because I'm guessing N is small enough for it not to matter and for there to be lots of other allocations happening in msbuild nearby, but I am getting curious.

Overall looks good. I think @wli3 makes a good suggestion to try to extract some smaller methods.

nguerrera commented:
I'm seeing this on a relatively recent build of dotnet new console
````
Target Performance Summary       
       14 ms  ResolveProjectReferences                   1 calls
       15 ms  _HandlePackageFileConflicts                1 calls
       20 ms  ResolveTargetingPacks                      1 calls
       21 ms  PrepareForBuild                            1 calls
       22 ms  ResolveAssemblyReferences                  1 calls
       22 ms  FindReferenceAssembliesForReferences       1 calls
       89 ms  ResolveFrameworkReferences                 1 calls

Task Performance Summary:
         8 ms  ResolvePackageAssets                       1 calls
       14 ms  ResolvePackageFileConflicts                1 calls
       16 ms  GetPackageDirectory                        2 calls
       22 ms  ResolveAssemblyReference                   1 calls
       88 ms  ResolveFrameworkReferences                 1 calls
```
I think we need to be more frugal in these tasks. I spent a lot of time optimizing ResolvePackageAssets. We have to be careful moving to something else that happens on incremental build for framework references.

We need to dig in. I am surprised that GetPackageDirectory is that slow. It hits the disk a bit, but do does ResolvePackageAssets. This could be related to nuget resolver slowness. ResolvePackageAssets cached all calls to the corresponding nuget API.

It is possible to do the same here. I will file a separate issue.

nguerrera commented:
I guess in fairness, ResolvePackageAssets isn't resolving anything. I will compare to 2.0.

nguerrera commented:
ReoslvePackageAssets is only 9ms for targeting 2.0.

ResolveFrameworkReferences is 43ms targeting 2.0 so not pay for play.

dsplaisted committed:
Add strings for duplicate framework references

dsplaisted committed:
Handle duplicate FrameworkReferences

Fixes #2890

dsplaisted committed:
Disable failing store test

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs:
```diff
@@ -46,7 +48,10 @@ protected override void ExecuteCore()
                 ITaskItem targetingPack;
                 if (!resolvedTargetingPacks.TryGetValue(frameworkReference.ItemSpec, out targetingPack))
                 {
-                    Log.LogError(Strings.UnknownFrameworkReference, frameworkReference.ItemSpec);
```
Do you mean the guideline "Do not have public members that can either throw or not throw exceptions based on some option."?

I don't consider our tasks public APIs, so I don't think this guideline applies.

dsplaisted committed:
Code review feedback

dsplaisted commented:
I've updated the `ResolveFrameworkReferences` task to short-circuit out if there are no `FrameworkReference` items, so it should be closer to pay-for-play now.

If we want the new tasks to be as fast as ResolvePackageAssets, we may need to cache them in the same way (or merge them into one task and cache it all in one file).

nguerrera commented:
I filed a bug, #2908 for the perf.

dsplaisted committed:
Fix tests with updated stage 0

dsplaisted committed:
Remove workaround for dotnet/core-setup#4978

dsplaisted committed:
Ignore .xml files in "runtime pack"

dsplaisted committed:
Various test fixes

- Show xunit help correctly
- Apply specified SDK version to perf tests
- Don't clone perf tests in design-time build
- Use DOTNET_ROOT instead of DOTNET_HOME

dsplaisted commented:
Tests I want to add as part of this:

- [ ] Disable targeting pack download (EnableTargetingPackDownload)
- [ ] Runtime Framework version selection
  - [ ] RuntimeFrameworkVersion metadata on FrameworkReference item
  - [ ] RuntimeFrameworkVersion MSBuild property
  - [ ] TargetLatestRuntimePatch metadata on FrameworkReference
  - [ ] TargetLatestRuntimePatch MSBuild property
- [ ] Targeting pack version selection
- [ ] NETSDK1082 - No runtime pack available
- [ ] NETSDK1083 - RuntimeIdentifier not recognized (see It_fails_for_unsupported_rid and It_allows_unsupported_rid_with_override publish tests)

dsplaisted committed:
Add link to bug for possibly removing dotnet store, add comment to targets file

dsplaisted committed:
Add tests covering runtime framework and other version selection, fix bugs

dsplaisted committed:
Fix bugs and add tests for RuntimeIdentifier not recognized or runtime pack not available

dsplaisted commented:
@nguerrera @wli3 I'm done with the changes I want to make for this PR, if you would like to take a look and give an official sign-off.

🤞 that all the tests still pass 😄

nguerrera commented:
I presume this should be retargeted to release/3.0.1xx branch.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateFrameworkReferences.cs:
```diff
@@ -0,0 +1,62 @@
+    public class CheckForDuplicateFrameworkReferences : TaskBase
+    {
+        [Required]
+        public ITaskItem[] FrameworkReferenceItems { get; set; }
```
Why not name this `FrameworkReferences`. Items feels redundant with the `ITaskItem[]` type.

dsplaisted replied:
It was to resolve the singular/plural discrepancy between `FrameworkReference` as an MSBuild item name and `FrameworkReferences` as a task property :-)

nguerrera replied:
Foos="@(Foo)" is common. I don't see issue.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs:
```diff
@@ -55,168 +51,140 @@ public class ResolveFrameworkReferences : TaskBase
-                .ToDictionary(kfr => kfr.Name);
+                .ToList();
+
+            var frameworkReferenceDict = FrameworkReferences.ToDictionary(fr => fr.ItemSpec);
```
nit: I'd pref `Map` over `Dict` abbreviation.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs:
```diff
@@ -55,168 +51,140 @@ public class ResolveFrameworkReferences : TaskBase
         {
-            var knownFrameworkReferences = KnownFrameworkReferences.Select(item => new KnownFrameworkReference(item))
+            //  Perf optimization: If there are no FrameworkReference items, then don't do anything
+            //  (This means that if you don't have any direct framework references, you won't get any transitive ones either
```
I'm not sure this needs to be true. I would expect the transitive FrameworkReferences to be raised from assets files before we get here. Is it worth stating presumed implications on a feature that doesn't exist yet?

dsplaisted replied:
This target runs before the assets file is read (it's the one that generates the list of packages that need to be downloaded).

nguerrera replied:
Ah, of course. The task name is a little ambiguous. I think of ResolveXReferences as something that turns abstract references into paths. Maybe we should rename it.

nguerrera replied:
I think the actual limitation in parentheses here is fine. I understand why it's needed now.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateFrameworkReferences.cs:
```diff
@@ -0,0 +1,62 @@
+
+        protected override void ExecuteCore()
+        {
+            var duplicateItems = FrameworkReferenceItems.GroupBy(i => i.ItemSpec, StringComparer.OrdinalIgnoreCase).Where(g => g.Count() > 1);
```
Let's put the early return on know framework references here too for consistency. Can't hurt to spare the allocations.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveAppHosts.cs:
```diff
@@ -0,0 +1,173 @@
+                    var packTargetFramework = NuGetFramework.Parse(appHostPack.GetMetadata("TargetFramework"));
+                    return packTargetFramework.Framework.Equals(TargetFrameworkIdentifier, StringComparison.OrdinalIgnoreCase) &&
+                        ResolveFrameworkReferences.NormalizeVersion(packTargetFramework.Version) ==
+                            ResolveFrameworkReferences.NormalizeVersion(new Version(TargetFrameworkVersion));
```
This normalization of right-hand-side can be hoisted.

nguerrera commented on src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToStoreAProjectWithDependencies.cs:
```diff
@@ -161,22 +161,23 @@ public void store_nativeonlyassets()
         }
 
-        [CoreMSBuildOnlyFact]
+        [CoreMSBuildOnlyFact(Skip = "https://github.com/dotnet/sdk/issues/2914")]
```
Is this test netcoreapp3.0 specifc?

nguerrera replied:
@livarcocc Are you ok with disabling the store tests while we make a call. I'm a bit worried about taking on debt if we get the call to keep it.

dsplaisted replied:
This one started failing for .NET Core 1.x.  I discussed disabling these tests with Livar, although I'm not sure I called out the fact that one of the failing ones wasn't for .NET Core 3.

nguerrera replied:
It's a little concerning that we broke 1.x with this change. I can't think of a reason why that would happen. I'd prefer that we investigate in case we're breaking other aspects of exisitng projects.

livarcocc replied:
I wasn't aware that we were breaking 1.x projects as well. I thought the breaking was for netcoreapp3.0 projects.

nguerrera commented on src/Tests/Microsoft.NET.PerformanceTests/BuildPerf.cs:
```diff
@@ -71,6 +71,7 @@ public void BuildWebLarge(ProjectPerfOperation operation)
             var testDir = _testAssetsManager.CreateTestDirectory("WebLarge", identifier: operation.ToString());
             Console.WriteLine($"Mirroring {sourceProject} to {testDir}...");
             FolderSnapshot.MirrorFiles(sourceProject, testDir.Path);
+            TestContext.Current.WriteGlobalJson(testDir.Path);
```
Why is this necessary?

dsplaisted replied:
So that if you run the perf tests with an `-sdkVersion` argument, it will respect that.  Most of the time the global.json will be written when the test directory is created, but in this case `MirrorFiles` wipes out anything in the target directory, so we have to explicitly write it again.

dsplaisted committed:
Apply code review feedback

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs:
```diff
@@ -55,168 +51,140 @@ public class ResolveFrameworkReferences : TaskBase
                     {
-                        foreach (var runtimePackNamePattern in knownFrameworkReference.RuntimePackNamePatterns.Split(';'))
+                        string runtimePackRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(
+                            new RuntimeGraphCache(this).GetRuntimeGraph(RuntimeGraphPath),
```
Can this be hoisted?

nguerrera replied:
I guess it's cached, but still.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs:
```diff
@@ -55,168 +51,140 @@ public class ResolveFrameworkReferences : TaskBase
+                        string runtimePackRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(
+                            new RuntimeGraphCache(this).GetRuntimeGraph(RuntimeGraphPath),
+                            RuntimeIdentifier,
+                            knownFrameworkReference.RuntimePackRuntimeIdentifiers.Split(';'),
```
This too.

nguerrera reviewed (commented):
After realizing which task is ResolveFrameworkReferences, I'm very curious what takes so much time. I filed a separate bug already, so you can ignore the perf questions on this pass if you want.

nguerrera commented:
I suspect the reading of the runtime graph.

dsplaisted commented:
> I suspect the reading of the runtime graph.

@nguerrera I added a separate task to read the runtime graph, and that task took 12 ms while `ResolveFrameworkReferences` still took 24 ms.  So it doesn't seem to be just the loading of the runtime graph that's taking up the time.

nguerrera commented:
See some of my comments on #2908. I'm also seeing a one-time init of FrameworkNameProvider being expensive. So far I haven't been able to profile accurately in a configuration that doesn't JIT nuget dlls. I think that's part of the answer here too.

nguerrera commented:
I'll move some more thoughts on to #2908. Let's use that for perf discussion. I think with that issue filed, you can ignore my perf comments that came in late here with respect to following up on this PR feedback.
